internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,442 @@
+			// next pathname in the queue.
+			writeBytesWithNull(h, []byte(filepath.ToSlash(referent)))
+			continue
+		}
+
+		// For both directories and regular files, we must create a file system
+		// handle in order to read their contents.
+		fh, err := os.Open(pathname)
+		if err != nil {
+			return nil, errors.Wrap(err, "cannot Open")
+		}
+
+		if fi.IsDir() {
+			childrenNames, err := sortedListOfDirectoryChildrenFromFileHandle(fh)
+			if err != nil {
+				_ = fh.Close() // already have an error reading directory; ignore Close result.
+				return nil, errors.Wrap(err, "cannot get list of directory children")
+			}
+			for _, childName := range childrenNames {
+				switch childName {
+				case ".", "..", "vendor", ".bzr", ".git", ".hg", ".svn":
+					// skip
+				default:
+					pathnameQueue = append(pathnameQueue, filepath.Join(pathname, childName))
+				}
+			}
+		} else {
+			written, err = io.Copy(h, newLineEndingReader(fh))            // fast copy of file contents to hash
+			err = errors.Wrap(err, "cannot Copy")                         // errors.Wrap only wraps non-nil, so elide guard condition
+			writeBytesWithNull(h, []byte(strconv.FormatInt(written, 10))) // format file size as base 10 integer
+		}
+
+		// Close the file handle to the open directory without masking possible
+		// previous error value.
+		if er := fh.Close(); err == nil {
+			err = errors.Wrap(er, "cannot Close")
+		}
+		if err != nil {
+			return nil, err // early termination iff error
+		}
+	}
+
+	return h.Sum(nil), nil
+}
+
+// lineEndingReader is a `io.Reader` that converts CRLF sequences to LF.
+//
+// Some VCS systems automatically convert LF line endings to CRLF on some OS
+// platforms. This would cause the a file checked out on those platforms to have
+// a different digest than the same file on platforms that do not perform this
+// translation. In order to ensure file contents normalize and hash the same,
+// this struct satisfies the io.Reader interface by providing a Read method that
+// modifies the file's contents when it is read, translating all CRLF sequences
+// to LF.
+type lineEndingReader struct {
+	src             io.Reader // source io.Reader from which this reads
+	prevReadEndedCR bool      // used to track whether final byte of previous Read was CR
+}
+
+// newLineEndingReader returns a new lineEndingReader that reads from the
+// specified source io.Reader.
+func newLineEndingReader(src io.Reader) *lineEndingReader {
+	return &lineEndingReader{src: src}
+}
+
+// Read consumes bytes from the structure's source io.Reader to fill the
+// specified slice of bytes. It converts all CRLF byte sequences to LF, and
+// handles cases where CR and LF straddle across two Read operations.
+func (f *lineEndingReader) Read(buf []byte) (int, error) {
+	buflen := len(buf)
+	if f.prevReadEndedCR {
+		// Read one less byte in case we need to insert CR in there
+		buflen--
+	}
+	nr, er := f.src.Read(buf[:buflen])
```

Comment:
That pattern was moddled after https://github.com/golang/go/blob/master/src/io/io.go#L392. When I first read the copied code (same BSD licence), the lack of a prompt error check also caught my eye, until I saw _why_ the author wrote it like that. The code does indeed check the error condition, but sometimes a read will return some bytes, and an error. In this case, when it read any bytes, the code still wants to do something with those bytes. The final error is returned to upstream client when the function returns.

The only catch is handling the case the case where the there were no bytes read on the source `io.Reader`, yet the `lineEndingReader` instance is sitting on a _CR_. In this case, line 202 emits one last successful read by returning one final _CR_ byte, and clearing the error if it was set. The upstream client gets another read with a single byte--not optimal, but correct--and we expect it would invoke Read another time. At that point the error returned by the underlying source `io.Reader` will be returned to the upstream client.